pym/bob/cmds/build.py

```diff
@@ -450,23 +459,83 @@ def _runShell(self, step, scriptName):
             print("# END BUILD SCRIPT", file=f)
         os.chmod(absRunFile, stat.S_IRWXU | stat.S_IRGRP | stat.S_IWGRP |
             stat.S_IROTH | stat.S_IWOTH)
-        cmdLine = ["/bin/bash", runFile, "__run"]
+        cmdLine = ["/bin/bash", runFile, "__runNoLog"]
         if self.__verbose < 0:
             cmdLine.append('-q')
         elif self.__verbose == 1:
             cmdLine.append('-v')
         elif self.__verbose >= 2:
             cmdLine.append('-vv')
 
-        proc = subprocess.Popen(cmdLine, cwd=step.getWorkspacePath(), env=runEnv)
+        logFile = open(os.path.join(step.getWorkspacePath(), '..', 'log.txt'), 'a')
+
+        oup = pty.openpty()
+        erp = pty.openpty()
+        pipe = os.pipe()
+
+        for p in pipe:
+            flag = fcntl.fcntl(p, fcntl.F_GETFL)
+            fcntl.fcntl(p, fcntl.F_SETFL, flag | os.O_NONBLOCK)
+
+        def child_exit(signum, frame):
+            try:
+                os.write(pipe[1], b"1")
+            except IOError as e:
+                if e.errno not in [errno.EAGAIN, errno.EINTR]:
+                    raise
+
+        signal.signal(signal.SIGCHLD, child_exit)
+
+        proc = subprocess.Popen(cmdLine, cwd=step.getWorkspacePath(), env=runEnv,
+                stdout=oup[1], stderr=erp[1])
+
+        ansi_escape = re.compile(r'\x1B\[([0-9]{1,2}(;[0-9]{1,2})?)?[m|K]')
+
         try:
+            done = False
+            while not done:
+                try:
+                    ready, _, _ = select.select([oup[0], erp[0], pipe[0]], [], [])
+                    for r in ready:
+                        if r == pipe[0]:
+                            os.read(pipe[0], 1)
+                            done = True
+                        else:
+                            oldLen = 0
+                            data = bytearray(b'')
+                            # read all available data from fd
+                            blockSize = 1024
+                            while True:
+                                data = data + os.read(r, blockSize)
```

Comment:
Don't you have to set the non-block flag for the pty file descriptors too? I guess these calls block until the client terminates... :thinking: